google/cloud/bigtable/internal/rate_limiter.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+
+#include "google/cloud/internal/clock.h"
+#include "google/cloud/log.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <cstdint>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A threadsafe interface for rate limiting.
+ *
+ * This class does not sleep. It is the responsibility of the caller to sleep.
+ * For example:
+ *
+ * @code
+ * auto clock = std::make_shared<internal::SteadyClock>();
+ * RateLimiter limiter(clock, rate);
+ * while (MoreThingsToDo()) {
+ *   auto wait = limiter.acquire(1);
+ *   std::this_thread_sleep_for(wait);
+ *   DoOneThing();
+ * }
+ * @endcode
+ *
+ * To allow for some smoothing, this class can bank permits. For example, if the
```

Comment:
Is there a max on this?

Reply:
Yeah, added a sentence to address the max.

Reply:
` The maximum stored permits is configurable by @p max_stored_permits`

Reply:
That `@p var` is a doxygen thing. It references a function parameter, so we wouldn't use it for the class documentation. It would belong on the constructor..... if I had documented that function..... :see_no_evil:

I think I deleted this section in favor of something coryan wrote anyways.

Reply:
ack sg